python/cudf/cudf/core/series.py

```diff
@@ -216,17 +216,25 @@ def __setitem__(self, key, value):
         if (
             not isinstance(
                 self._frame._column.dtype,
-                (cudf.core.dtypes.DecimalDtype, cudf.CategoricalDtype),
+                (
+                    # Casting only for string and non-decimal numeric
+                    # columns
+                    cudf.core.dtypes.DecimalDtype,
+                    cudf.CategoricalDtype,
+                    cudf.ListDtype,
+                    cudf.StructDtype,
+                ),
             )
             and hasattr(value, "dtype")
             and _is_non_decimal_numeric_dtype(value.dtype)
         ):
             # normalize types if necessary:
-            if not is_integer(key):
-                to_dtype = np.result_type(
-                    value.dtype, self._frame._column.dtype
-                )
-                value = value.astype(to_dtype)
+            # In contrast to Column.__setitem__ (which downcasts the value to
+            # the dtype of the column) here we upcast the series to the
+            # larger data type mimicing pandas
```

Comment:
```suggestion
            # larger data type mimicking pandas
```

Reply:
I think I’m asking a different question. This code appears to permit a string column to be assigned an int value (a string column is not disallowed by the excluded dtypes for decimal, categorical, list, struct). However, the opposite seems to not be allowed because the value must be “non-decimal numeric.” Is that sensible? Why this asymmetry? Can those paths be added to tests?